ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java

```diff
@@ -162,18 +307,52 @@ public boolean equals(Object o) {
     if (!version.equals(blobId.version)) {
       return false;
     }
-    if (!partitionId.equals(blobId.partitionId)) {
+    if (!flag.equals(blobId.flag)) {
       return false;
     }
-    if (!uuid.equals(blobId.uuid)) {
+    if (!datacenterId.equals(blobId.datacenterId)) {
       return false;
     }
-
-    return true;
+    if (!accountId.equals(blobId.accountId)) {
+      return false;
+    }
+    if (!containerId.equals(blobId.containerId)) {
+      return false;
+    }
+    if (!partitionId.equals(blobId.partitionId)) {
+      return false;
+    }
+    return uuid.equals(blobId.uuid);
   }
 
   @Override
   public int hashCode() {
     return Utils.hashcode(new Object[]{version, partitionId, uuid});
   }
+
+  /**
+   * Populate the missing fields with legacy values, and determine the version of format to serialize. Only
+   * when all the four fields of {@code flag}, {@code datacenterId}, {@code accountId}, and {@code containerId}
+   * are available, the serialization format version will be set to {@link #BLOB_ID_V2}, otherwise it will be
+   * {@link #BLOB_ID_V1}.
+   */
+  private void populateMissingFieldsAndDetermineVersion() {
+    version = flag == null && accountId == null && containerId == null && datacenterId == null ? BlobId.BLOB_ID_V1
```

Comment:
Thanks for comments from both of you, and it seems a common point is that we should be careful when enabling the write path. Let me clarify two things:
1. after this pr, both blob id v1 and v2 can be understood;
2. this pr does not enable write blob id v2. Write is actually enabled by router depending on how router uses the `BlobIdBuilder`. In the original version of this pr, when router supplies one of the four required fields by v2 (`flag`, `datacenterId`, `accountId`, and `containerId`), it will enable v2.

Based on the second point, router has the control to write v2. However, as discussed with Priyesh, we want to make this safe and unambiguous, so now a `CURRENT_VERSION` is introduced into `BlobId`. When this is set to v1, it will construct blobId v1 regardless how we set the four fields required by v2. Once we decide to write v2 at router side, we also need to change `CURRENT_VERSION` to v2. This will gives us more confidence.

Because of this change, I have rewritten a significant part of the unite tests, because the logic of deciding version is not necessary to test.